### PR TITLE
Image Download/Provision Support

### PR DESCRIPTION
**Pull request by pathuang-sia**
Adding the following new properties:
Cloud agent level:
- gallery image: CloudInitDataSource, AzureGalleryImageProperties {sasURI, version (last part of urn)}
- virtual machine image: CloudInitDataSource

Node agent level: 
- virtual hard disk:  CloudInitDataSource, AzureGalleryImageProperties {sasURI, version (last part of urn)}
- virtual machine: CloudInitDataSource

common: 
- definitions for  CloudInitDataSource and DownloadStatus under existing property Status

### CODE REVIEW

File: rpc/common/moc_common_common.proto [hunk 1]
```diff
@@ -74,8 +81,9 @@ message Version {
 message Status {
 	Health health = 1;
 	ProvisionStatus provisioningStatus = 2;
-	Error lastError = 3;
-	Version version = 4;
+	DownloadStatus downloadStatus = 3;
+	Error lastError = 4;
```
Comment (manoj-kadam) [inline rpc/common/moc_common_common.proto:85]: I don't think we should change existing numbering in here for backward compatibility

File: rpc/cloudagent/compute/galleryimage/moc_cloudagent_galleryimage.proto [hunk 2]
```diff
@@ -59,6 +59,8 @@ message GalleryImage {
 	string sourcePath = 11;
 	Tags tags = 12;
 	ImageSource sourceType = 13;
+	string version = 14;
+	CloudInitDataSource cloudInitDataSource = 16;
```
Comment (madhanrm) [inline rpc/cloudagent/compute/galleryimage/moc_cloudagent_galleryimage.proto:63]: I dont see this in the design md

Comment (pathuang-sia) [reply rpc/cloudagent/compute/galleryimage/moc_cloudagent_galleryimage.proto:63]: the version is in vm-download.md and the cloudInitDataSource is in the vm-provision.md file

File: rpc/common/moc_common_common.proto [hunk 3]
```diff
@@ -76,6 +84,7 @@ message Status {
 	ProvisionStatus provisioningStatus = 2;
 	Error lastError = 3;
 	Version version = 4;
+	DownloadStatus downloadStatus = 5;
```
Comment (madhanrm) [inline rpc/common/moc_common_common.proto:87]: Try to include DownloadStatus within Status.

Comment (madhanrm) [inline rpc/common/moc_common_common.proto:87]: Try to include DownloadStatus within Status.

Comment (pathuang-sia) [reply rpc/common/moc_common_common.proto:87]: Do you mean to deconstruct the subproperties of downloadStatus into status?

Comment (madhanrm) [reply rpc/common/moc_common_common.proto:87]: You can pack DownloadStatus into Status as an element. Try to pack it in a way so you can unpack it.
There may be other elements in status, so add one for DownloadStatus. This wouldnt give it a struct, but more like arbitrary data

Comment (pathuang-sia) [reply rpc/common/moc_common_common.proto:87]: Isn't this already the case? The DownloadStatus is part of Status right now

File: rpc/nodeagent/storage/virtualharddisk/moc_nodeagent_virtualharddisk.proto [hunk 4]
```diff
@@ -69,6 +69,7 @@ message VirtualHardDisk {
 	Entity entity = 18;
 	Tags tags = 19;
 	ImageSource sourceType = 20;
+	CloudInitDataSource cloudInitDataSource = 22;
```
Comment (madhanrm) [inline rpc/nodeagent/storage/virtualharddisk/moc_nodeagent_virtualharddisk.proto:72]: Is this to copy over the properties from GalleryImage to VHD on vmimage creation

File: rpc/common/moc_common_common.proto [hunk 5]
```diff
@@ -55,13 +55,21 @@ enum HealthState {
 message Error {
 	string Message = 1;
 	int32 Code = 2;
+	string Parameters = 3;
```
Comment (madhanrm) [inline rpc/common/moc_common_common.proto:58]: what will this hold?

**Review by madhanrm — APPROVED** [review]

File: rpc/common/moc_common_common.proto (hunk 5 above)
Comment (pathuang-sia) [reply rpc/common/moc_common_common.proto:58]: comma separated, key value pairs which define parameters and values optionally encountered in the message, as defined by the idc team

**Comment by madhanrm** [comment]
Good to merge

**Review by manoj-kadam — APPROVED** [review]

**Pull request merged**